### PR TITLE
This is another implementation correspond to #5931

### PR DESCRIPTION
**Pull request by yeweiasia**
fix snippet suggestion always displayed all items

Signed-off-by: yewei <yeweiasia@gmail.com>

#### What it does
This is another implementation correspond to https://github.com/theia-ide/theia/pull/5931

#### Why to rewrite
1.~~the code to find out all  lineOffsets in the line make no sense in case of snippet code suggestion, https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/contrib/snippets/browser/snippetCompletionProvider.ts#L86-L117~~
2.implement the same feature with less line of code, that means less maintain works, meanwhile sync to vscode is easy if  function name is kept to be unchanged.
3.improve the snippet suggestion filter and replacement logic
### case 1: suggestion when trigger in the word
vscode: the pattern is the prefix text before the cursor
trigger_in_vscode

us: the pattern is the whole word, so more Irrelevant will be filtered
trigger_in_theia

### case 2: trigger in the word and select a item
vscode: only replace the prefix text
replacement_in_vscode

vertiacl_line_replacement_vscode

us: replace the whole word
replacement_in_theia

vertiacl_line_replacement_theia

#### How to test

1.theia 0.9.0 with [redhat.java-0.46.0.vsix](https://github.com/redhat-developer/vscode-java/releases/download/v0.46.0/redhat.java-0.46.0.vsix)
2.open a maven project and edit a java file
3.type keyword like "if".
image
4.select "ifelse" or other snippet suggestion.
5.without this patch whole suggested item body will be append to "if"  which is demonstrated in https://github.com/theia-ide/theia/issues/5926
autocompletion1
6.with this patch "if" will  replaced by suggested item body.
snippetcodecompletion

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
Why not simply merge the complete solution #5931 rather than this alternative?

**Comment by yeweiasia** [comment]
> Why not simply merge the complete solution #5931 rather than this alternative?

this is a complete solution as well, copy code from vscode is easy but the implementation of vscode  is confusing and therefore it will take more time to understand what this code is doing. I think change to our own implementation is easier to locate problems and I will keep this feature in sync with vscode.

**Comment by akosyakov** [comment]
@yeweiasia Did you use git blame to find why they came up with such logic and tested that corresponding issues from VS Code repo and reproducible with your changes? There are probably reasons even if we don't see them now.

For example here they reverted an attempt to simplify logic: https://github.com/microsoft/vscode/commit/6a5b2d68894c3f1c0a6f03e71cffe8b4c001ac6c

**Comment by akosyakov** [comment]
@vince-fugnitto @BroKun could you test please this PR and if it is good please approve and merge.

@BroKun Could you also verify that your changes in https://github.com/theia-ide/theia/pull/5883 works nicely with this PR as a base?

**Comment by yeweiasia** [comment]
> @yeweiasia Did you use git blame to find why they came up with such logic and tested that corresponding issues from VS Code repo and reproducible with your changes? There are probably reasons even if we don't see them now.
> 
> For example here they reverted an attempt to simplify logic: [microsoft/vscode@6a5b2d6](https://github.com/microsoft/vscode/commit/6a5b2d68894c3f1c0a6f03e71cffe8b4c001ac6c)

@akosyakov I've test https://github.com/microsoft/vscode/issues/58807
trigger_more_than_100_characters

File: packages/monaco/src/browser/monaco-snippet-suggest-provider.ts [hunk 1]
```diff
@@ -24,17 +24,32 @@ export class MonacoSnippetSuggestProvider implements monaco.modes.ISuggestSuppor
...
+    async provideCompletionItems(model: monaco.editor.ITextModel, position: monaco.Position, context: monaco.modes.SuggestContext): Promise<monaco.modes.ISuggestResult> {
         const languageId = model.getModeId(); // TODO: look up a language id at the position
         await this.loadSnippets(languageId);
-        const suggestions = this.snippets.get(languageId) || [];
+        const snippetsForLanguage = this.snippets.get(languageId) || [];
+        const suggestions: MonacoSnippetSuggestion[] = [];
+        const word = model.getWordAtPosition(position);
+        const positionEndsInEmptyChar = model.getLineContent(position.lineNumber).substr(0, position.column - 1).match(/\s$/);
+        snippetsForLanguage.forEach(snippet => {
+            if (word && this.isPatternInText(word.word, snippet.prefix, true)) {
```
Comment (BroKun) [inline packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: Just checking the current word will lose some of the matches. Snippet's prefix can be multi words.

Comment (yeweiasia) [reply packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: @BroKun thanks for reporting this, Could you show me more information about your multi words test case?

Comment (BroKun) [reply packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: import-single
import-multi

Comment (yeweiasia) [reply packages/monaco/src/browser/monaco-snippet-suggest-provider.ts]: @BroKun thanks, I will take this case nto consideration

**Comment by vince-fugnitto** [comment]
With the latest implementation I'm wondering if it's still worthwhile in providing an implementation which differs from that of VSCode https://github.com/theia-ide/theia/pull/5931. The previous argument that there is less lines of code and that it is simpler does not seem to be valid. I can see resemblance (inspiration perhaps) between the latest implementation and that of VSCode.

**Comment by yeweiasia** [comment]
> With the latest implementation I'm wondering if it's still worthwhile in providing an implementation which differs from that of VSCode #5931. The previous argument that there is less lines of code and that it is simpler does not seem to be valid. I can see resemblance (inspiration perhaps) between the latest implementation and that of VSCode.

yep, I missed the multi word match case and it seems unnecessary to provide this implementation. I will add the replacement logic to #5931, keep the main code block sync with vscode seems benefit more.

**Comment by yeweiasia** [comment]
#5931 is instead of this PR

**Pull request closed without merging**